### PR TITLE
kubeadm: Split discovery from JoinConfiguration

### PR DESCRIPTION
**Pull request by rosti**
**What this PR does / why we need it**:

This change splits out discovery fields from JoinConfiguration by performing
the following changes:

- Introduce a BootstrapTokenDiscovery structure, that houses configuration
  options needed for bootstrap token based discovery.

- Introduce a FileDiscovery structure, that houses configuration options
  (currently only a single option) needed for KubeConfig based discovery.

- Introduce a Discovery structure, that houses common options (such as
  discovery timeout and TLS bootstrap token) as well as pointer to an instance
  of either BootstrapTokenDiscovery or FileDiscovery structures.

- Replace the old discovery related JoinConfiguration members with a single
  Discovery member.

This change is required in order to cleanup the code of unnecessary logic and
make the serialized JoinConfiguration more structured (and therefore, more
intuitive).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
refs kubernetes/kubeadm#911, refs kubernetes/kubeadm#963

**Special notes for your reviewer**:

This is another one of those big and somewhat hard to review PRs, that are the stepping stones in the path to `v1beta1`.

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews
/area kubeadm
/kind api-change
/kind enhancement
/assign @luxas
/assign @timothysc
/assign @fabriziopandini

**Release note**:

```release-note
kubeadm: JoinConfiguration now houses the discovery options in a nested Discovery structure, which in turn has a couple of other nested structures to house more specific options (BootstrapTokenDiscovery and FileDiscovery)
```

### CODE REVIEW

File: cmd/kubeadm/app/cmd/join.go
Comment (dixudx) [inline cmd/kubeadm/app/cmd/join.go]: Default will be nil.
Remove it.

File: cmd/kubeadm/app/cmd/join.go
Comment (dixudx) [inline cmd/kubeadm/app/cmd/join.go]: Ditto.

File: cmd/kubeadm/app/util/config/testdata/validation/invalid_nodecfg.yaml
Comment (dixudx) [inline cmd/kubeadm/app/util/config/testdata/validation/invalid_nodecfg.yaml]: Missing newline?

File: cmd/kubeadm/app/util/config/testdata/conversion/node/internal.yaml
Comment (dixudx) [inline cmd/kubeadm/app/util/config/testdata/conversion/node/internal.yaml]: This will break old configurations.

For coming `v1beta1`, we cound use a new section like this.

But for current `v1alpha3`, shall we use below instead in `types.go`,
```go
Discovery Discovery `json:",inline"`
```

File: cmd/kubeadm/app/util/config/testdata/conversion/node/v1alpha3.yaml
Comment (dixudx) [inline cmd/kubeadm/app/util/config/testdata/conversion/node/v1alpha3.yaml]: Ditto for those following yaml files.

Better not change the struction of current released v1alpha3. We could do this in coming v1beta1 instead.

**Review by dixudx — CHANGES_REQUESTED** [review]
For current `v1alpha3`, I'd prefer not to update the json structure. Use `inline` instead.

We could update that in next coming `v1beta1`.

File: cmd/kubeadm/app/util/config/testdata/conversion/node/internal.yaml
Comment (rosti) [reply cmd/kubeadm/app/util/config/testdata/conversion/node/internal.yaml]: Actually, it won't, since `v1alpha3` is also unreleased and is massively different than `v1alpha2` of 1.11.
I am also not a fan of inlining it, because that way the context of the options in `Discovery` is gone.

**Comment by rosti** [comment]
>For current v1alpha3, I'd prefer not to update the json structure. Use inline instead.
>
>We could update that in next coming v1beta1.

This looks like a double work for me and I am willing to actually hold of this PR for `v1beta1` if we are going to split it into user and back facing interface changes.
I am not a big fan of inlining the Discovery structure either - it gives us more context that way and makes the config file more structured.
The fact is that this change is literally tiny compared to the rest if the changes we have done in `v1alpha3`. I don't believe that a user facing interface change here is going to break anyone, since `v1alpha3` is unreleased.

**Review by timothysc — COMMENTED** [review]
/hold 

please coordinate this with @fabriziopandini there are overlapping issues right now.

**Comment by rosti** [comment]
Rebased the PR on v1beta1

@fabriziopandini @timothysc PTAL

**Comment by timothysc** [comment]
@rosti bot says it still needs a rebase. 
I'd like to chat about config changes tomorrow during office hours too.

**Review by timothysc — COMMENTED** [review]
Given that you and @fabriziopandini have been working on this one I'll defer to him on lgtm 

/approve 
/assign @fabriziopandini

**Review by fabriziopandini — COMMENTED** [review]
@rosti thank you for this PR and +100 for keeping this going across two cycles!

Overall this PR looks fine, but the things that confused a little bit me is the move from tree token fields (`discoveryToken`, `tlsBootstrapToken` and `token` to override the former two) to two token fields. Some other minors inline

File: cmd/kubeadm/app/apis/kubeadm/types.go
Comment (rosti) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: How about `KubeConfigFile`? Yet it sounds kind of the same to me.

**Comment by rosti** [comment]
Fixed the `--token` flag semantics (of `kubeadm join`).

File: cmd/kubeadm/app/apis/kubeadm/types.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: Ok according to original Lucas design

File: cmd/kubeadm/app/apis/kubeadm/types.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: Ok according to original Lucas design

File: cmd/kubeadm/app/apis/kubeadm/types.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: Leave as it is ATM

File: cmd/kubeadm/app/apis/kubeadm/v1alpha3/conversion.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/apis/kubeadm/v1alpha3/conversion.go]: ok according to original Lucas design/the semantic of --token acting as a proxy for TLSBootstrapToken/DiscoveryToken

File: cmd/kubeadm/app/apis/kubeadm/v1beta1/defaults.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/apis/kubeadm/v1beta1/defaults.go]: ok according to original Lucas design

File: cmd/kubeadm/app/apis/kubeadm/types.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: Please make this string; we are not accepting more than one value ATM and there is no plan for supporting >1 API Endpoints as far as I know

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: Please rename to  `ValidateDiscoveryBootstrapToken`

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (fabriziopandini) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: This should go away when `APIServerEndpoints` changes from `[]string` to `string`

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: kindly rename `ValidateJoinDiscoveryTokenAPIServer` into `ValidateDiscoveryTokenAPIServer`for consistency

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: Please rename to `ValidateDiscoveryFile `

File: cmd/kubeadm/app/cmd/join.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/join.go]: Please fix the flag documentation explaining the semantics of this flag

**Review by fabriziopandini — CHANGES_REQUESTED** [review]
@rosti
/approve
Please fix minors still pending and then ready for lgtm

After your explanation I added a note on all my previous comments that does not apply anymore, and make it clear where IMO you should fix minors

File: cmd/kubeadm/app/cmd/join.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/join.go]: I'm not fans of this solution but we can leave with it ATM

File: cmd/kubeadm/app/apis/kubeadm/types.go
Comment (rosti) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: This will come in as a followup PR as the change is not trivial (currently there is a lot of code in kubeadm, that looks upon this as a `[]string` and doing correct handling of multiple API Servers).

**Comment by rosti** [comment]
A few minor details to address in a couple of hours and this will be ready to go.

/hold

**Pull request merged**
